x/metadata/types/scope.go

```diff
@@ -461,21 +457,38 @@ func (p Party) ValidateBasic() error {
 	return nil
 }
 
+// ValidatePartiesAreUnique makes sure that no two provided parties are the same.
+func ValidatePartiesAreUnique(parties []Party) error {
```

Comment:
I always have to stop and think about the slice ranges. And I'm used to the i = 0 to 1-n, j = i+1 to n format, so it was easier for me to write it that way and be sure it's not off-by-one somewhere.

A little added reason is that parties are `[]Party`, a slice of concrete (non-reference) values. But `.IsSameAs` needs a reference. The `lParty` and `rParty` variables are re-used for each iteration which can lead to problems if you're storing a reference to them outside the loop. It's not a concern in this case since `.IsSameAs` doesn't hold onto the provided arg, but the `gosec` linter would complain, requiring a nolint comment (as seen in one or two other places).

But really, I feel safer with the old-school for loop iteration than with sub-slices since I'm more familiar with the former when it comes to comparing values in a single list/slice/array.

Reply:
Heh. So, I accidentally checked in a version of `ValidatePartiesAreUnique` that had both the original loops and this:
```golang
	for i, pi := range parties[:len(parties)-1] {
		for j := range parties[i+1:] {
			if pi.IsSameAs(&parties[j]) {
				return fmt.Errorf("duplicate parties not allowed: address = %s, role = %s, indexes: %d, %d",
					parties[i].Address, parties[i].Role, i, j)
			}
		}
	}
```

That new version has some problems that the original didn't:
1. It panics if `parties` is nil.
2. It panics if `parties` is empty.
3. The `j` index is not the same here as it is in the other one. It starts at 0 and goes to len(parties)-1-i. E.g., with three entries, when i = 0, j = 0, 1, then i = 1, j = 0. So that `&parties[j]` is not the correct entry to be comparing.

Basically, it only ever works right when there's exactly 1 entry in `parties`.

Here is what it'd look like after fixing it:
```golang
	if len(parties) <= 1 {
		return nil
	}
	for i, pi := range parties[:len(parties)-1] {
		for jj := range parties[i+1:] {
			j := i + 1 + jj
			if pi.IsSameAs(&parties[j]) {
				return fmt.Errorf("duplicate parties not allowed: address = %s, role = %s, indexes: %d, %d",
					parties[i].Address, parties[i].Role, i, j)
			}
		}
	}
```

I kind of like the original version better though, so I'm going to keep it like that.

Relatedly, I realized that I forgot to add unit tests on it. So I'll add some.